File: Time Measurement/PHP/execute.php
<?php

namespace execute;

require_once("lib.php");

$d=null;
$o=null;
$e=null;
$r=null;

$res="";
$ec=0;

function main($rd) {
	global $d,$o,$e,$r,$res,$ec;
	$d=$rd;
	$o=co2f($d->out,STDOUT);
	$e=co2f($d->err,STDERR);
	$r=ro2f($d->result);

	switch ($d->multiple) {
		case MMNone:   single(); break;
		case MMSerial: serial(); break;
		case MMSpawn:  spawn();  break;
	}

	fwrite($r,$res);
	closeIO($o);
	closeIO($e);
	closeIO($r);
	exit($ec==-1?1:$ec);
}

function single() {
	global $d,$res,$ec;
	$p=new SP($d->command);

	$st=hrtime(true);
	$p->run();
	$en=hrtime(true);

	$t=descTime($en-$st);
	$e=$p->descEC();
	$res=clean(<<< "Result"
		time: $t
		process id: $p->pid
		$e
	Result).PHP_EOL;
	$ec=$p->ec;
}

function serial() {
	global $d,$res,$ec;
	$pl=SP::multiple($d->command);
	$lp=$pl[count($pl)-1];

	$st=hrtime(true);
	foreach ($pl as $p) {
		$p->run();
		if ($p->ec!=0) {
			$lp=$p;
			break;
		}
	}
	$en=hrtime(true);

	$res=implode(PHP_EOL,[
		"time: ".descTime($en-$st),
		...array_map(function($p){ return "process$p->order id: ".($p->pid<0?"N/A":$p->pid); },$pl),
		$lp->descEC(),""
	]);
	$res="time: ".descTime($en-$st).PHP_EOL;
	foreach ($pl as $p) $res.="process$p->order id: ".($p->pid<0?"N/A":$p->pid).PHP_EOL;
	$res.=$lp->descEC().PHP_EOL;

	$ec=$lp->ec;
}

function spawn() {
	global $d;
	$pl=SP::multiple($d->command);

	$st=hrtime(true);
	foreach ($pl as $p) $p->start();
	foreach ($pl as $p) $p->wait();
	$en=hrtime(true);

	SP::collect($pl,$st,$en);
}

class SP {
	private $r;
	private $s;
	private $args;
	public $order=0;
	public $pid=-1;
	public $ec=0;

	public function __construct($args) {
		$this->args=$args;
	}

	public static function multiple($commands) {
		$n=1;
		$l=[];
		foreach ($commands as $c) {
			$p=new SP($c);
			$p->order=$n;
			array_push($l,$p);
			$n++;
		}
		return $l;
	}

	public static function collect($pl,$st,$en) {
		global $res,$ec;
		$ec=0;

		$r=["time: ".descTime($en-$st)];
		foreach ($pl as $p) {
			if ($p->ec>$ec) $ec=$p->ec;
			array_push($r,"process$p->order id: $p->pid",$p->descEC());
		}
		array_push($r,"");
		$res=implode(PHP_EOL,$r);
	}

	public function start() {
		global $o,$e;
		try {
			$this->r=proc_open($this->args,[0=>STDIN,1=>$o,2=>$e],$p);
			if ($this->r==false) throw Exception();
		}
		catch(Exception $e) {
			$as=implode(" ",$this->args);
			error("実行に失敗しました: $as");
		}
	}

	public function wait() {
		$s=proc_get_status($this->r);
		$ec=proc_close($this->r);
		if (!$s["running"]) $ec=$s["exitcode"];
		$this->pid=$s["pid"];
		$this->ec=$ec;
		$this->s=$s;
	}

	public function run() {
		$this->start();
		$this->wait();
	}

	public function descEC() {
		$s=$this->s;
		return $s["signaled"] ? "terminated due to signal ".strval($s["termsig"]) : "exit code: ".strval($this->ec);
	}

}



function co2f($d,$inherit) {
	switch ($d) {
		case "inherit": return $inherit;
		case "discard": return ["file","/dev/null","w"];
		default: return $this->fh($d);
	}
}

function ro2f($d) {
	switch ($d) {
		case "stdout": return STDOUT;
		case "stderr": return STDERR;
		default: return $this->fh($d);
	}
}

$opened=[];
function fh($path) {
	if (array_key_exists($path,$this->opened)) return $this->opened[$path];
	try {
		$f=fopen($path,"a");
		$this->opened[$path]=$f;
		return $f;
	}
	catch(Exception $e){ error("指定したパスには書き込みできません: ".$path); }
}

function closeIO($fh) {
	if (is_resource($fh)) fclose($fh);
}



function descTime($nSec) {
	$t="";
	$r=$nSec/(3600*1e+9);$v=floor($r);
	if ($v>=1) $t.="${v}h ";
	$r=($r-$v)*60;$v=floor($r);
	if ($v>=1) $t.="${v}m ";
	$r=($r-$v)*60;$v=floor($r);
	if ($v>=1) $t.="${v}s ";
	$r=($r-$v)*1000;
	$t.=sprintf("%07.3fms",$r);
	return $t;
}

function descEC($s) {
	return $s["signaled"] ? "terminated due to signal ".strval($s["termsig"]) : "exit code: ".strval($s["exitcode"]);
}

File: Time Measurement/Julia/execute.jl
using .lib

module execute
	export exec
	import ..Data,..clean,..exitWithError,..MMNone,..MMSerial,..MMSpawn,..MMThread
	using Printf,Base.Threads

	d=nothing
	o=nothing
	e=nothing
	r=nothing

	ec=0
	res=""

	function exec(rd::Data)
		global d,o,e,r
		d=rd
		o=co2f(d.out,stdout)
		e=co2f(d.err,stderr)
		r=ro2f(d.result)

		if d.multiple==MMNone single()
		elseif d.multiple==MMSerial serial()
		elseif d.multiple==MMSpawn spawn()
		elseif d.multiple==MMThread thread()
		end

		print(r,res)
		close(o)
		close(e)
		close(r)
		exit(ec)
	end

	function single()
		global res

		p=SP(`$(d.command)`,join(d.command," "))

		sec=@elapsed begin
			runSP(p)
		end

		res=clean("""
			time: $(descTime(sec))
			process id: $(p.pid)
			$(descEC(p))
		""")
		ec=p.ec
	end

	function serial()
		global res

		pl=multiple()
		lp=pl[length(pl)]

		sec=@elapsed begin
			for p in pl
				runSP(p)
				if p.ec!=0
					lp=p
					break
				end
			end
		end

		res=join([
			"time: $(descTime(sec))",
			map(p->"process$(p.order) id: $(p.pid<0 ? "N/A" : p.pid)",pl)...,
			descEC(lp),""
		],"\n")
		ec=lp.ec
	end

	function spawn()
		pl=multiple()

		sec=@elapsed begin
			for p in pl startSP(p) end
			for p in pl waitSP(p) end
		end

		collect(pl,sec)
	end

	function thread()
		pl=multiple()

		sec=@elapsed begin
			@threads for p in pl
				runSP(p)
			end
		end

		collect(pl,sec)
	end

	Base.@kwdef mutable struct SP
		cmd::Base.CmdRedirect
		description::String
		process=nothing
		order::Int = 0
		pid::Int32 = -1
		ec::Int = 0
		SP(cmdObj,desc) = new(pipeline(cmdObj,stdout=stdout,stderr=stderr,stdin=stdin),desc)
	end
	function startSP(p::SP)
		try
			p.process=run(p.cmd,wait=false)
			pid=getpid(p.process)
			p.pid=pid
		catch e
			exitWithError("実行に失敗しました: $(p.description)")
		end
	end
	function waitSP(p::SP)
		wait(p.process)
		p.ec=p.process.exitcode
	end
	function runSP(p::SP)
		startSP(p)
		waitSP(p)
	end
	function descEC(p::SP)::String
		return @sprintf("exit code: %.d",p.ec)
	end
	function multiple()::Array{SP,1}
		pl=fill(-1,length(d.command))
		n=1
		l::Array{SP,1}=[]
		for c in d.command
			if occursin(r"#\{\}\(\)\[\]<>\|\&\*\?~;",c)
				sh=get(ENV,"SHELL","sh")
				p=SP(`$sh -c $c`)
			else
				s=c
				s=replace(s,"\\"=>"\\\\")
				s=replace(s,"`"=>"\\`")
				s=replace(s,"\$"=>"\\\$")
				try
					p=SP(eval(Meta.parse("`$s`")),d.command)
				catch e
					exitWithError("パースに失敗しました: $c")
				end
			end
			p.order=n
			n+=1
			append!(l,[p])
		end
		return l
	end
	function collect(pl::Array{SP,1},sec)
		global res,ec

		res=join(Iterators.flatten([
			["time $(descTime(sec))"],
			map(function (p)
			if p.ec>ec ec=p.ec end
				return ["process$(p.order) id: $(p.pid)",descEC(p)]
			end,pl)...,[""]
		]),"\n")
	end

	function co2f(d,i)
		if d=="inherit" return i
		elseif d=="discard" return devnull
		else return fh(d)
		end
	end

	function ro2f(d)
		if d=="stdout" return stdout
		elseif d=="stderr" return stderr
		else return fh(d)
		end
	end

	opened=Dict()
	function fh(path)
		if haskey(opened,path) return opened[path]
		else
			try
				io=open(path,"a")
				opened[path]=io
				return io
			catch e
				exitWithError("指定したパスには書き込みできません: "+path)
			end
		end
	end

	function descTime(sec)::String
		t=""
		r=sec/3600
		v=floor(r)
		if v>=1 t*=@sprintf("%.0fh ",v) end
		r=(r-v)*60
		v=floor(r)
		if v>=1 t*=@sprintf("%.0fm ",v) end
		r=(r-v)*60
		v=floor(r)
		if v>=1 t*=@sprintf("%.0fs ",v) end
		r=(r-v)*1000
		t*=@sprintf("%07.3fms",r)
		return t
	end

end